src/linalg/impl_linalg.rs

```diff
@@ -1140,3 +1140,379 @@ where A: LinalgScalar
+                "tensordot: cannot contract over {} axes; a.ndim()={}, b.ndim()={}",
+                n,
+                nda,
+                ndb
+            );
+            ((nda - n)..nda).zip(0..n).map(|(ia, ib)| (ia, ib)).unzip()
+        }
+        AxisSpec::Pair(aa, bb) => {
+            assert_eq!(
+                aa.len(),
+                bb.len(),
+                "tensordot: axes length mismatch (a has {}, b has {})",
+                aa.len(),
+                bb.len()
+            );
+            (aa, bb)
+        }
+    };
+
+    // Normalise negative indices
+    for ax in &mut axes_a {
+        if *ax < 0 {
+            *ax += nda;
+        }
+    }
+    for ax in &mut axes_b {
+        if *ax < 0 {
+            *ax += ndb;
+        }
+    }
+
+    // Validate
+    for &ax in &axes_a {
+        assert!(
+            (0..nda).contains(&ax),
+            "tensordot: axis {} out of bounds for a (ndim={})",
+            ax,
+            nda
+        );
+    }
+    for &ax in &axes_b {
+        assert!(
+            (0..ndb).contains(&ax),
+            "tensordot: axis {} out of bounds for b (ndim={})",
+            ax,
+            ndb
+        );
+    }
+
+    // Shape checks
+    for (ia, ib) in axes_a.iter().zip(&axes_b) {
+        let da = a.shape()[*ia as usize];
+        let db = b.shape()[*ib as usize];
+        assert_eq!(
+            da, db,
+            "tensordot: shape mismatch along contraction axis: a[{}]={} vs b[{}]={}",
+            ia, da, ib, db
+        );
+    }
+
+    // Determine non-contracted axes
+    let notin_a: Vec<usize> = (0..nda as usize)
+        .filter(|k| !axes_a.iter().any(|&ax| ax as usize == *k))
+        .collect();
+    let notin_b: Vec<usize> = (0..ndb as usize)
+        .filter(|k| !axes_b.iter().any(|&ax| ax as usize == *k))
+        .collect();
+
+    // Reorder axes
+    let mut newaxes_a = notin_a.clone();
+    newaxes_a.extend(axes_a.iter().map(|&x| x as usize));
+    let mut newaxes_b = axes_b.iter().map(|&x| x as usize).collect::<Vec<_>>();
+    newaxes_b.extend(notin_b.iter().copied());
+
+    // Matrix shapes
+    let m = notin_a.iter().fold(1, |p, &ax| p * a.shape()[ax]);
+    let k = axes_a.iter().fold(1, |p, &ax| p * a.shape()[ax as usize]);
+    let n = notin_b.iter().fold(1, |p, &ax| p * b.shape()[ax]);
+
+    let a_dyn = a.view().into_dimensionality::<IxDyn>().unwrap();
+    let b_dyn = b.view().into_dimensionality::<IxDyn>().unwrap();
+
+    let a_perm = a_dyn.permuted_axes(IxDyn(&newaxes_a));
+    let a_std = a_perm.as_standard_layout();
+
+    let b_perm = b_dyn.permuted_axes(IxDyn(&newaxes_b));
+    let b_std = b_perm.as_standard_layout();
+
+    let a2 = a_std
+        .into_shape_with_order(Ix2(m, k))
+        .expect("reshaping a to 2D");
+    let b2 = b_std
+        .into_shape_with_order(Ix2(k, n))
+        .expect("reshaping b to 2D");
+
+    let c2 = a2.dot(&b2);
+
+    let mut out_shape: Vec<usize> = notin_a.iter().map(|&ax| a.shape()[ax]).collect();
```

Comment:
You're allocating 3 vectors in this function. Can't you re-use one of them for `out_shape`?